exporter/exporter.go

```diff
@@ -57,8 +59,106 @@ resource "lightstep_metric_dashboard" "exported_dashboard" {
 }
 `
 
-func escapeSpanQuery(input string) string {
-	return strings.Replace(input, "\"", "\\\"", -1)
+const unifiedDashboardTemplate = `
+resource "lightstep_dashboard" "exported_dashboard" {
+  project_name = var.project
+  dashboard_name = "{{.Attributes.Name}}"
+{{range .Attributes.Charts}}
+  chart {
+    name = "{{.Title}}"
+    rank = "{{.Rank}}"
+    type = "{{.ChartType}}"
+{{range .MetricQueries}}
+    query {
+      query_name          = "{{.Name}}"
+      display             = "{{.Display}}"
+      hidden              = {{.Hidden}}
+      query_string        = {{escapeQueryString .TQLQuery}}
+	}
+{{end}}
+  }
+{{end}}
+}
+`
+
+func escapeHCLString(input string) string {
+	input = strings.Replace(input, "\"", "\\\"", -1)
+	input = strings.Replace(input, "\n", "\\\t", -1)
+	input = strings.Replace(input, "\r", "\\\t", -1)
+	input = strings.Replace(input, "\t", "\\\t", -1)
+	input = strings.Replace(input, "\\", "\\\\", -1)
+	return input
+}
+
+func escapeQueryString(input string) string {
+	// Use "heredoc" syntax if the query contains any newlines or other characters that'd
+	// need to be escaped and make the single line representation less convenient to work with.
+	if strings.Contains(input, "\"") ||
+		strings.Contains(input, "\n") ||
+		strings.Contains(input, "\r") ||
+		strings.Contains(input, "\t") ||
+		strings.Contains(input, "\\") {
+		return "<<EOT\n" + input + "\nEOT"
+	} else {
+		return escapeHCLString(input)
+	}
+}
+
+// dashboardUsesQueryString returns true if any chart in the dashboard
+// uses the query string format.
+//
+// The Provider does not support mixing legacy query charts and
+// query-string charts in the same dashboard and will return an error
+// if there is a mix.
+func dashboardUsesQueryString(d *client.UnifiedDashboard) (bool, error) {
+
+	hasQueryString := false
+	hasLegacyQuery := false
+	for _, chart := range d.Attributes.Charts {
+		for _, q := range chart.MetricQueries {
+			if len(q.TQLQuery) > 0 {
+				hasQueryString = true
+			} else {
+				hasLegacyQuery = true
+			}
+			if hasLegacyQuery && hasQueryString {
+				return false, fmt.Errorf("dashboards containing a mix of legacy and query string charts are not supported")
+			}
+		}
+	}
+	if hasLegacyQuery {
+		return false, nil
+	}
+	return true, nil
+}
+
+func exportToHCL(wr io.Writer, d *client.UnifiedDashboard) error {
```

Comment:
Split this worker function out of `Run` so the export could be unit tested more easily. This also adds logic to check which resource type should be used based on the kinds of queries in the charts.